cachi2/core/package_managers/npm.py

```diff
@@ -173,6 +188,120 @@ def get_sbom_components(self) -> list[dict[str, str]]:
         packages = self._dependencies if self.lockfile_version == 1 else self._packages
         return [{"name": package.name, "version": package.version} for package in packages]
 
+    def get_dependencies_to_download(self) -> Dict[str, Dict[str, Any]]:
+        """Return a list of URL dependencies to download."""
+        packages = self._dependencies if self.lockfile_version == 1 else self._packages
+        return {
+            package.name: {
+                "version": package.version,
+                "url": package.resolved_url,
+                "integrity": package.integrity,
+            }
+            for package in packages
+            if "file:" not in package.resolved_url
+        }
+
+
+def _get_external_requirement_filepath(download_dir: RootedPath, requirement_url: str) -> Path:
+    """Verify that the file matches at least one of the expected checksums.
+
+    Note: any checksum algorithms not supported by python's hashlib will be skipped.
+
+    If none of the checksums match, log all the mismatches and skipped algorithms at WARNING level,
+    then raise an exception.
+
+    :param download_dir: Output folder where dependencies will be downloaded
+    :param requirement_url: Resolved URL for package
+    :raises ValueError: when passed requirement_url is neither `url` nor `vcs`
+    """
+    parsed_url = urlparse(requirement_url)
+    namespace = parsed_url.path.removeprefix("/").split("/")[0]
+    repo = parsed_url.path.removeprefix("/").split("/")[-1].removesuffix(".git")
+    ref = parsed_url.fragment
+
+    filepath = Path(
+        download_dir,
+        parsed_url.hostname,  # host
+        namespace,
+        repo,
+        f"{repo}-external-gitcommit-{ref}.tgz",
+    )
+    return filepath
+
+
+def clone_repo_pack_archive(
+    url: str,
+    download_path: Path,
+) -> None:
+    """
+    Clone a repository and pack its content as tar.
+
+    :param str url: URL for file download
+    :param str download_path: Output folder where dependencies will be downloaded
+    :raise FetchError: If download failed
+    """
+    parse_url = url.removeprefix("git+").split("#")
+    repo_url = parse_url[0]
+    repo_commit = parse_url[1] if len(parse_url) > 1 else None
+
+    # Create missing directories
+    directory = os.path.dirname(download_path)
+    os.makedirs(directory, exist_ok=True)
+    clone_as_tarball(str(repo_url), str(repo_commit), download_path)
+
+
+def get_dependencies(download_dir: RootedPath, deps_to_download: Dict[str, Dict[str, Any]]) -> Dict:
+    """
+    Download npm dependencies.
+
+    Receives the url (proxy_repo_url), the destination directory (download_dir)
+    and the dependencies to be downloaded (deps_to_download).
+
+    :param Path download_dir: Path to download file to.
+    :param list[str] deps_to_download: Dict of dependencies to be downloaded.
+    :return: Dict of downloaded tarballs from npm registries for integrity check.
+    """
```

Comment:
Docstring is not quite right
* there's no proxy_repo_url
* the types (IMO types don't belong in the docstring anyway)

Reply:
Docstring updated